authorization/role/service/role_management_service.go

```diff
@@ -38,3 +47,96 @@ func (r *RoleManagementServiceImpl) ListByResource(ctx context.Context, resource
+
+		rt, err := r.repo.ResourceRepository().Load(ctx, resourceID)
+
+		if err != nil {
+			return err
+		}
+
+		roleRef, err := r.repo.RoleRepository().Lookup(ctx, roleName, rt.ResourceType.Name)
+		if err != nil {
+			return err
+		}
+
+		ir := rolerepo.IdentityRole{
+			ResourceID: resourceID,
+			IdentityID: identityID,
+			RoleID:     roleRef.RoleID,
+		}
+
+		return r.repo.IdentityRoleRepository().Create(ctx, &ir)
+	})
+
+	return err
+}
+
+// Assign assigns an identity ( users or organizations or teams or groups ) with a role, for a specific resource
+// IMPORTANT: This is a transactional method, which manages its own transaction/s internally
+func (r *RoleManagementServiceImpl) Assign(ctx context.Context, assignedBy uuid.UUID, identitiesToBeAssigned []uuid.UUID, resourceID string, roleName string) error {
+	// check if the current user token belongs to a user who has the necessary privileges
+	// for assigning roles to other users.
+
+	permissionService := permservice.NewPermissionService(r.repo)
+	hasScope, err := permissionService.HasScope(ctx, assignedBy, resourceID, authorization.ManageTeamsInSpaceScope)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"resource_id": resourceID,
+			"identity_id": assignedBy,
+			"role":        roleName,
+		}, "error determining if user has manage scope")
+		return err
+	}
+	if !hasScope {
+		log.Error(ctx, map[string]interface{}{
+			"resource_id": resourceID,
+			"identity_id": assignedBy,
+			"role":        roleName,
+		}, "user not authorizied to assign roles")
+		return errors.NewForbiddenError("user is not authorized to assign roles")
+	}
+
+	// In batch assignment of roles all selected users must have previously been assigned
+	// privileges for the resource, otherwise the invitation workflow should be used instead
+	for _, identityIDAsUUID := range identitiesToBeAssigned {
+		assignedRoles, err := r.repo.IdentityRoleRepository().FindIdentityRolesByIdentityAndResource(ctx, resourceID, identityIDAsUUID)
+		if err != nil {
+			log.Error(ctx, map[string]interface{}{
+				"resource_id": resourceID,
+				"identity_id": assignedBy,
+				"role":        roleName,
+			}, "error looking up existing assignments")
+			return err
+		}
+		if len(assignedRoles) == 0 {
+			log.Error(ctx, map[string]interface{}{
+				"resource_id": resourceID,
+				"identity_id": identityIDAsUUID,
+				"role":        roleName,
+			}, "identity not part of a resource cannot be assigned a role")
+			return errors.NewBadParameterErrorFromString("identityID", identityIDAsUUID, fmt.Sprintf("cannot update roles for an identity %s without an existing role", identityIDAsUUID))
+		}
+	}
+
+	// Now that we have confirmed that all users have pre-existing role assignments
+	// we can proceed with the assignment of roles.
+	for _, identityIDAsUUID := range identitiesToBeAssigned {
+		err = r.assign(ctx, identityIDAsUUID, resourceID, roleName)
```

Comment:
You wish to wrap each role assignment in its own transaction?  What if for some reason one of the transactions fails?  Do we wish to support a "partial" success?

Reply:
Initially I did think of having it in one transaction ( that's how the tests are now ) - then yesterday thought of partial success. But now I realize that its a bad idea. Among many other things, it'll be hard to report errors to users while also letting them know of the partial success.